### PR TITLE
Modify heuristic for dummy group selection

### LINKED ISSUE #1016: Disconnected dummy atoms in some cases with complete_rings=False
E.g. methane → cyclobutadiene

With `complete_rings=False`, we generate a mapping like:
methane-cyclobutane-atom-map

Due to the heuristic [here](https://github.com/proteneer/timemachine/blob/03b9dd1e8bb86f06291c941840563a66287475f2/timemachine/fe/dummy.py#L253), we only generate a single dummy group for B containing all 4 dummy atoms. Then, turning off all but one of the anchor bonds to maintain factorizability results in a disconnected dummy atom.

Potential fix: modify to use two passes:

1. Initialize dummy groups as connected components of the subgraph induced by dummy atoms, then 

2. merge groups with a common anchor. In the above example, this would result in 2 dummy groups with no disconnected atoms.

### PR DESCRIPTION
**Pull request by mcwitt**
### Summary

* Adds test ensuring that the alchemical hybrid is fully connected for the methane -> cyclobutadiene example in #1016 
* Modifies heuristic to ensure dummy atoms are always connected
* Adds an assertion to `single_topology.setup_end_state` checking that the hybrid molecule has a single connected component

### Todo

* [x] Run vacuum benchmarks
* [ ] How does typical number of dummy groups change with the new heuristic?

### Background

The existing heuristic for dummy group identification ([here](https://github.com/proteneer/timemachine/blob/03b9dd1e8bb86f06291c941840563a66287475f2/timemachine/fe/dummy.py#L242-L261))

1. Builds a subgraph of mol_b induced by the dummy atoms
2. Augments the induced subgraph with extra edges connecting atoms that are distance 2 apart. I.e. if (i, j) and (j, k) are edges in the induced subgraph, also adds (i, k)
3. Identifies connected components of the augmented induced subgraph with dummy groups

The motivation for step 2 is to allow dummy groups to be merged if they share an anchor atom, even if the original dummy groups do not have bond interactions.

However, this can lead to disconnected dummy atoms in some cases where we have 3 or more disconnected components of the induced subgraph that are connected in the augmented induced subgraph (i.e. by step 2) through 2 or more different anchor atoms (see the example in #1016).

This PR modifies the heuristic to

1. Build a subgraph of mol_b induced by the dummy atoms (same as before)
2. Find connected components of the (non-augmented) induced subgraph
3. Arbitrarily choose an anchor for each connected component
4. Merge components with identical anchors

This is similar to the existing heuristic with the important difference that it cannot result in disconnected dummy atoms (by construction).

Like the existing heuristic, we make an arbitrary choice of anchor atom for a given dummy group when there are several possibilities.

Unlike the existing heuristic, the approach here does not necessarily merge dummy groups that are separated by a single core atom. If this is desirable, we should be able to implement it in the present approach by ranking candidates in order of increasing number of dummy groups before making an arbitrary selection.

Fixes #1016

### CODE REVIEW

File: timemachine/fe/dummy.py [hunk 1]
```diff
@@ -1,151 +1,179 @@
...
 
-    ```
-    mol = Chem.MolFromSmiles("FC1CC1(F)N")
-    core = [1, 2, 3]
-    dg = identify_dummy_groups(mol, core)
-    assert_set_equality(dg, [{0}, {4, 5}])
-    ```
+    Example
+    -------
+    TODO
```
Comment (mcwitt) [inline timemachine/fe/dummy.py:49]: TODO

File: attic/single_topology/dummy.py [hunk 2]
```diff
@@ -59,6 +59,73 @@ def identify_anchor_groups(bond_idxs, core, root_anchor):
     return layer_1_groups, layer_2_groups, layer_3_groups
 
 
+def identify_root_anchors(bond_idxs, core, dummy_atom):
```
Comment (mcwitt) [inline attic/single_topology/dummy.py:62]: On the fence about attic'ing vs. deleting this function

File: timemachine/fe/dummy.py [hunk 3]
```diff
@@ -1,151 +1,202 @@
...
+        canonicalize_bond(tuple(mapping[idx] for idx in bond)) for bond in bonds if all(idx in mapping for idx in bond)
+    ]
+
+
+_K = TypeVar("_K")
+_V = TypeVar("_V")
+
+
+def union_by_key(ts: Iterable[Tuple[_K, FrozenSet[_V]]]) -> Dict[_K, FrozenSet[_V]]:
+    d: DefaultDict[_K, FrozenSet[_V]] = defaultdict(frozenset)
```
Comment (jkausrelay) [inline timemachine/fe/dummy.py:199]: nit: add a one line doc string here

File: timemachine/fe/dummy.py [hunk 4]
```diff
@@ -1,151 +1,202 @@
...
+
+    def get_angle_anchors(bond_anchor):
+        valid_angle_anchors = [
+            angle_anchor
+            for angle_anchor in [n for n in bond_graph_b.neighbors(bond_anchor) if n in core_atoms_b]
+            if canonicalize_bond((bond_anchor, angle_anchor)) in core_bonds_b
+        ]
+        return valid_angle_anchors or [None]
+
+    anchored_dummy_group_assignments = (
```
Comment (jkausrelay) [inline timemachine/fe/dummy.py:144]: nit: Can you add a short comment? The 4 levels of for loops here took me a bit to parse.

File: tests/test_single_topology.py [hunk 5]
```diff
@@ -175,9 +170,29 @@ def testing_find_dummy_groups_and_multiple_anchors():
...
 
     with pytest.warns(MultipleAnchorWarning):
-        dgs, jks = single_topology.find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b)
-        assert dgs == [{0}]
-        assert jks == [(1, 2)]
+        dgs = single_topology.find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b)
+        assert dgs == {1: (2, {0})}
+
+
+def test_ethane_cyclobutane():
```
Comment (jkausrelay) [inline tests/test_single_topology.py:177]: nit: This is cyclobutadiene, cyclobutane is C1CCC1

File: tests/test_single_topology.py [hunk 6]
```diff
@@ -175,9 +170,29 @@ def testing_find_dummy_groups_and_multiple_anchors():
...
+    core = np.array([[2, 0], [4, 2], [0, 3], [3, 7]])
+    ff = Forcefield.load_default()
+    st = SingleTopology(mol_a, mol_b, core, ff)
+
+    g = nx.Graph()
+    g.add_nodes_from(range(st.get_num_atoms()))
+    g.add_edges_from(st.src_system.bond.potential.idxs)
+
+    # bond graph should be connected (i.e. no floating bits)
+    assert len(list(nx.connected_components(g))) == 1
```
Comment (jkausrelay) [inline tests/test_single_topology.py:195]: Should this check just be added to the SingleTopology init?

File: timemachine/fe/dummy.py [hunk 7]
```diff
@@ -1,151 +1,202 @@
...
-    2) Generate an induced graph using edges from above set.
-    3) Disconnected components of the resulting induced graph are our dummy groups.
+    While the choices of dummy groups is arbitrary, this function partitions the dummy atoms into multiple dummy groups
+    using a heuristic:
 
-    Code example:
+    1) Generate an induced subgraph containing only dummy atoms
+    2) Identify connected components of the induced subgraph
+    3) For each connected component, generate all possible choices of (bond) anchor atom
+    4) Merge connected components with the same bond anchor; this gives a mapping of anchor atom to dummy group for each
```
Comment (proteneer) [inline timemachine/fe/dummy.py:35]: missing a step: between 3) and 4) - the choice of the bond anchor that a dummy group belongs to is currently done randomly?

File: timemachine/fe/dummy.py [hunk 8]
```diff
@@ -1,151 +1,202 @@
...
+) -> FrozenSet[Tuple[int, int]]:
+    """Returns core-core bonds that are present in both mol_a and mol_b"""
+    bonds_a = get_romol_bonds(mol_a)
+    bonds_b = get_romol_bonds(mol_b)
+    a_to_c = {a: c for c, a in enumerate(core_atoms_a)}
+    b_to_c = {b: c for c, b in enumerate(core_atoms_b)}
+    return frozenset(translate_bonds(bonds_a, a_to_c)).intersection(frozenset(translate_bonds(bonds_b, b_to_c)))
+
+
+def translate_bonds(bonds: Collection[Tuple[int, int]], mapping: Dict[int, int]):
```
Comment (proteneer) [inline timemachine/fe/dummy.py:188]: nit: docstring

File: timemachine/fe/dummy.py (hunk 7 above)
Comment (mcwitt) [reply timemachine/fe/dummy.py:35]: It is not done randomly; the returned iterator will generate the outcomes for all possible choices. Will try to clarify here.

File: timemachine/fe/dummy.py (hunk 4 above)
Comment (mcwitt) [reply timemachine/fe/dummy.py:146]: Added in https://github.com/proteneer/timemachine/pull/1019/commits/8c5087628184388d8b8db2da24d0f6d522e8c853

File: tests/test_single_topology.py (hunk 6 above)
Comment (mcwitt) [reply tests/test_single_topology.py:195]: Good call, added this in https://github.com/proteneer/timemachine/pull/1019/commits/afbf684951fb5af969ceaaeb399a69d03344760d

File: tests/test_single_topology.py (hunk 5 above)
Comment (mcwitt) [reply tests/test_single_topology.py]: Fixed in https://github.com/proteneer/timemachine/pull/1019/commits/f2cdeaaf1f22b948c4e97d9bfbbc829e98499818

File: timemachine/fe/dummy.py (hunk 7 above)
Comment (mcwitt) [reply timemachine/fe/dummy.py]: Attempted to clarify in https://github.com/proteneer/timemachine/pull/1019/commits/daf51795268dac55e11595665deb7d51e2fd8566

File: timemachine/fe/dummy.py (hunk 8 above)
Comment (mcwitt) [reply timemachine/fe/dummy.py:190]: Added in https://github.com/proteneer/timemachine/pull/1019/commits/285bdc753592abbe038415f7a3041d39cc7c7feb

File: timemachine/fe/dummy.py [hunk 9]
```diff
@@ -1,151 +1,202 @@
...
-
-    1) Identify all 1-2, 1-3 edges between dummy atoms.
-    2) Generate an induced graph using edges from above set.
-    3) Disconnected components of the resulting induced graph are our dummy groups.
+    While the choices of dummy groups is arbitrary, this function partitions the dummy atoms into multiple dummy groups
+    using a heuristic:
 
-    Code example:
+    1) Generate an induced subgraph containing only dummy atoms
+    2) Identify connected components of the induced subgraph
```
Comment (proteneer) [inline timemachine/fe/dummy.py:33]: maybe we should make a note that this step forms the "basis" for our dummy groups, and that we never subdivide a basis group (i.e. we can only combine the basis into larger dummy groups, but never shrink them).

File: tests/test_dummy.py [hunk 10]
```diff
@@ -122,34 +63,33 @@ def test_identify_dummy_groups():
       0-------1
 
     """
-    bond_idxs = get_bond_idxs(Chem.MolFromSmiles("FC1CC1(F)N"))
+
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply tests/test_dummy.py:56]: Good call; strengthened the test in https://github.com/proteneer/timemachine/pull/1019/commits/6ccb51d7e5886e73ec14a3feccd995555e3ceb6d to check for the exact expected dummy groups and anchors.

[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_dummy.py:56]: Thanks - and if possible, one more base-case sanity test for an empty `core=[]` (just to make sure nested collections etc. behave as they should).

File: timemachine/fe/dummy.py (hunk 9 above)
Comment (mcwitt) [reply timemachine/fe/dummy.py]: Added a note on this in https://github.com/proteneer/timemachine/pull/1019/commits/ac5504e8a49bc5a4f97583985a7fa9373f1c7fb9

File: tests/test_dummy.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mcwitt) [reply tests/test_dummy.py:56]: Added in https://github.com/proteneer/timemachine/pull/1019/commits/9276d3368b207e75dd695cf4f70253e606f0faca

**Review by proteneer — APPROVED** [review]
lgtm!

**Review by jkausrelay — APPROVED** [review]
